temporal-sdk/src/main/java/io/temporal/worker/tuning/JVMSystemResourceInfo.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.worker.tuning;
+
+import com.sun.management.OperatingSystemMXBean;
+import io.temporal.common.Experimental;
+import java.lang.management.ManagementFactory;
+import java.time.Instant;
+import java.util.concurrent.locks.ReentrantLock;
+
+/** {@link SystemResourceInfo} implementation that uses JVM-specific APIs to get resource usage. */
+@Experimental
+public class JVMSystemResourceInfo implements SystemResourceInfo {
+  // As of relatively recent Java versions (including backports), this class will properly deal with
+  // containerized environments as well as running on bare metal.
+  // See https://bugs.openjdk.org/browse/JDK-8226575 for more details on which versions the fixes
+  // have been backported to.
+  OperatingSystemMXBean osBean = ManagementFactory.getPlatformMXBean(OperatingSystemMXBean.class);
+
+  private final ReentrantLock refreshLock = new ReentrantLock();
+  volatile Instant lastRefresh = Instant.EPOCH;
+  volatile double lastMemUsage = Double.NaN;
+  volatile double lastCpuUsage = Double.NaN;
+
+  @Override
+  public double getCPUUsagePercent() {
+    refresh();
+    return lastCpuUsage;
+  }
+
+  @Override
+  public double getMemoryUsagePercent() {
+    refresh();
+    return lastMemUsage;
+  }
+
+  @SuppressWarnings("deprecation") // deprecated APIs needed since replacements are for Java 14+
+  private void refresh() {
```

Comment:
Usually I'd just put this behind `synchronized`, I don't think a reentrant lock is what is needed (not sure how same thread would call back in here). And I think you should update everything all at once to avoid cases where two at the same time cause two refreshes (because you are not checking staleness inside the lock only outside). Can simplify to:

```java
private SystemInfo lastSystemInfo;

@Override
public double getCPUUsagePercent() {
    getSystemInfo().cpuUsagePercent;
}

@Override
public double getMemoryUsagePercent() {
    getSystemInfo().memoryUsagePercent;
}

private synchronized SystemInfo getSystemInfo() {
    if lastSystemInfo == null || Instant.now().isAfter(refreshed.plusMillis(100)) {     
        // Do logic
    }
    return lastSystemInfo;
}

private static class SystemInfo {
    private final Instant refreshed;
    private final double cpuUsagePercent;
    private final double memoryUsagePercent;

    private SystemInfo(Instant refreshed, double cpuUsagePercent, double memoryUsagePercent) {
        this.refreshed = refreshed;
        this.cpuUsagePercent = cpuUsagePercent;
        this.memoryUsagePercent= memoryUsagePercent;
    }
}
```

Just reads a bit more thread safe to me

Reply:
My concern is that I'm not sure what is going on within the JVM calls and I didn't want to encounter a thread pinning issue like Quinn had brought up before because of something happening inside of those, hence wanting to avoid `sychronized`. Reentrant lock just because that's seemingly used everywhere in the codebase regardless of need, not because I actually care about reentrancy.

The little info class makes sense to me though.

I had a check inside the lock, too, to avoid double-refreshing and oddly it was still causing me issues, but I'll make sure that works after using the class which ought to help.

Reply:
If You'll get the same issue with `ReentrantLock` as you will with `synchronized` for this case. If you're not using any of the more advanced methods on `ReentrantLock` and you are not calling it recursively, it is the same as `synchronized`.

Reply:
This is the pinning thing I'm referring to: https://mikemybytes.com/2024/02/28/curiosities-of-java-virtual-threads-pinning-with-synchronized/ which was the reason to use it over `synchronized`. I agree that the locking semantics aren't any different.

Reply:
Ah virtual threads, yes (though I am guessing the internals of the getting of the system info stuff doesn't yield to other threads, but unsure).

Reply:
Yeah we should avoid performing IO under `synchronized ` , if it doesn't yield it doesn't matter

Reply:
I sort of doubt it would too, but, regardless may as well be safe